### PR TITLE
Fix evictions test

### LINKED ISSUE #39412: go vet check failed
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

go vet check
src\github.com\kubernetes\kubernetes\test\integration\evictions\eviction
s_test.go:214: unreachable code
src\github.com\kubernetes\kubernetes\test\integration\evictions\eviction
s_test.go:188: result of fmt.Errorf call not used

**Kubernetes version** (use `kubectl version`):

**Environment**:
- **Cloud provider or hardware configuration**:
- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Install tools**:
- **Others**:

**What happened**:

**What you expected to happen**:

**How to reproduce it** (as minimally and precisely as possible):

**Anything else do we need to know**:

### PR DESCRIPTION
**Pull request by mengqiy**
**What this PR does / why we need it**:
Fixes bugs in evictions test. Make vet happy.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #39412

Ref: #39452

cc: @calebamiles

### CODE REVIEW

File: test/integration/evictions/evictions_test.go
Comment (calebamiles) [inline test/integration/evictions/evictions_test.go]: you may want to test that `doneCount` matches a previously set constant to make the test slightly less brittle.

File: test/integration/evictions/evictions_test.go
Comment (calebamiles) [inline test/integration/evictions/evictions_test.go]: might it be possible to check that the eviction has been completed inside of the goroutines spawned on line 177. The variable name `doneCh` may be considered a bit misleading given that we haven't checked that the pod was actually evicted before saying that we're done. If the final check that a pod was actually evicted was moved into the goroutine where we requested the pod be evicted this for loop could possibly be removed

File: test/integration/evictions/evictions_test.go
Comment (calebamiles) [inline test/integration/evictions/evictions_test.go]: does this call not produce an error?

File: test/integration/evictions/evictions_test.go
Comment (calebamiles) [inline test/integration/evictions/evictions_test.go]: it might be slightly cleaner (a little less cognitive overhead) to use a `switch` statement rather than testing the `if` statements

File: test/integration/evictions/evictions_test.go
Comment (calebamiles) [inline test/integration/evictions/evictions_test.go]: Is returning from the goroutine after receiving a conflict error the correct thing to do? nothing will be sent over `doneCh` in that case and wouldn't we fall through to the timeout case on line 213? Five minutes might be a long time to wait before moving on with the rest of the test. Have you collected any rough estimates on how long you expect evicting 10 pods to take when there isn't an error? I am aware that using channels might be the more "canonical" design pattern to communicate the status of the goroutine  (e.g. error encountered/no problems) but it might be simpler to use a `sync.WaitGroup` and make each goroutine responsible for quitting after the default timeout. A buffered `errCh` could allow you to store errors concurrently and process them later.

File: test/integration/evictions/evictions_test.go
Comment (calebamiles) [inline test/integration/evictions/evictions_test.go]: is there any reason to not call `t.Fatalf` here or are you thinking you'd like to be able to move this block around in the future?

**Review by calebamiles — CHANGES_REQUESTED** [review]
Looks generally pretty good but a slightly less canonical but easier (less use of channels and loop labels) might make the test more maintainable in the future but that's pure speculation on my part.

**Comment by ymqytw** [comment]
@calebamiles Thanks for your review.
I'm using `sync.WaitGroup` and refactored the test based on your suggestion. PTAL

**Comment by ymqytw** [comment]
@k8s-bot unit test this

**Comment by ymqytw** [comment]
@k8s-bot unit test this

File: test/integration/evictions/evictions_test.go
Comment (calebamiles) [inline test/integration/evictions/evictions_test.go]: you could defer these statements when creating the channels and resource management might be a little easier to follow

Comment (calebamiles) [inline test/integration/evictions/evictions_test.go]: we could defer these statements when creating the channels and resource management might be a little easier to follow

File: test/integration/evictions/evictions_test.go
Comment (calebamiles) [inline test/integration/evictions/evictions_test.go]: it doesn't look like we ever send `false` over `doneCh` so it's probably fine to increment immediately without the check. Also we _could_ make `doneCh` of type struct since we don't do anything with the value going over the channel. Since we adopted a shared variable pattern already (with the sync.WaitGroup) you could have a shared counter that you update via `atomic.AddUint32` or similar so this could be cleaned up slightly since you could simply check the value of the counter at the end without ranging over the channel that doesn't provide us a lot of useful information

File: test/integration/evictions/evictions_test.go
Comment (calebamiles) [inline test/integration/evictions/evictions_test.go]: due to the length of this conditional block (due to the length of the anonymous function) it might be a little easier to follow along if we assign `err` on line 186 and send the error over `errCh` if it's non `nil` and then we can bring the lines in the `else` block if we return out of the goroutine after sending a non `nil` error over `errCh`

File: test/integration/evictions/evictions_test.go
Comment (calebamiles) [inline test/integration/evictions/evictions_test.go]: if we return from the goroutine here can we pull out the `e = clientSet.Core().Pods(ns.Name).Delete(podName, deleteOption)` to after the `switch` statement so as to not duplicate the line since we handle both the `default` and `err == nil` cases the same way by deleting the pod and collecting any error returned

**Review by calebamiles — CHANGES_REQUESTED** [review]
things look good to me. a few small possible nits if you'd like to make the changes. If you're entirely happy with things as they are then shoot me a comment and I'll approve. Mostly these are my (probably outdated) thoughts about readability that I admit are fairly subjective. Thanks for making the previous changes and caring so much about testibility!

**Comment by calebamiles** [comment]
made a few small reviews @ymqytw, thanks for making the previous changes so quickly!

File: test/integration/evictions/evictions_test.go
Comment (ymqytw) [inline test/integration/evictions/evictions_test.go]: I'm not following you. Can you give me a snippet?

File: test/integration/evictions/evictions_test.go
Comment (ymqytw) [inline test/integration/evictions/evictions_test.go]: Done.

File: test/integration/evictions/evictions_test.go
Comment (ymqytw) [inline test/integration/evictions/evictions_test.go]: I have removed `doneCh`. But we need to close `errCh` before we iterate over it. Otherwise it will keep waiting on this channel.

File: test/integration/evictions/evictions_test.go
Comment (ymqytw) [inline test/integration/evictions/evictions_test.go]: Done.

**Comment by ymqytw** [comment]
@calebamiles I addressed 3 comments. 1 left for your further explanation.

**Comment by ymqytw** [comment]
@k8s-bot unit test this

**Comment by ymqytw** [comment]
@k8s-bot unit test this

File: test/integration/evictions/evictions_test.go
Comment (calebamiles) [inline test/integration/evictions/evictions_test.go]: here you go @ymqytw https://gist.github.com/calebamiles/8e62c22a7f82c915f0bb5779240b16e8

**Review by calebamiles — CHANGES_REQUESTED** [review]
added a snippet to explain my last comment which also shows the idea of using `atomic` to update an evicted pods counter rather than using a channel to collect done signals since we're only going to range over the channel at the end of the test anyway. if you don't like the ideas let me know in comment because you have LGTM otherwise

File: test/integration/evictions/evictions_test.go
Comment (calebamiles) [inline test/integration/evictions/evictions_test.go]: we may want to define this variable closer to where it's used and possibly use `var doneCount int32 = 0` to avoid the explicit cast. Also a `uint32` is probably fine

File: test/integration/evictions/evictions_test.go
Comment (calebamiles) [inline test/integration/evictions/evictions_test.go]: `errCh` may need to be at worst `3*numOfEvictions` I believe

**Review by calebamiles — CHANGES_REQUESTED** [review]

File: test/integration/evictions/evictions_test.go
Comment (ymqytw) [inline test/integration/evictions/evictions_test.go]: Done.

File: test/integration/evictions/evictions_test.go
Comment (ymqytw) [inline test/integration/evictions/evictions_test.go]: Yes, after using the code in your snippet, `errCh` needs `3*numOfEvictions`.

**Comment by ymqytw** [comment]
@calebamiles Thanks for your snippet. I updated the code. PTAL
I will squash the commits when they look good.

**Review by calebamiles — APPROVED** [review]
The only nit I would make is that `func TestConcurrentEvictionRequests` could probably be the first thing declared in the file after the `const` block  with the idea being that a test file should have the test first and supporting functions declared after so it's easiest for humans to understand but LGTM, @ymqytw

**Comment by calebamiles** [comment]
@fejta do you want to add the `LGTM` label on the PR after squash or should I?

**Comment by ymqytw** [comment]
@calebamiles Reorder the functions in the test and add a comment for the test. And then squashed. PTAL

**Comment by calebamiles** [comment]
looks great @ymqytw. lgtm

**Comment by ymqytw** [comment]
@k8s-bot gce etcd3 e2e test this

**Pull request merged**
